iree/task/executor_test.cc

```diff
@@ -0,0 +1,181 @@
+
+namespace {
+
+static thread_local volatile uint64_t xxx = 0;
+
+static void simulate_work(iree_task_tile_context_t* tile_context) {
+  iree_prng_splitmix64_state_t state;
+  iree_prng_splitmix64_initialize(xxx, &state);
+  bool slow = false;  // tile_context->workgroup_xyz[0] % 3 == 1;
+  if (tile_context->workgroup_xyz[0] == 128 ||
+      tile_context->workgroup_xyz[0] == 1023) {
+    // Introduce big variance to highlight work stealing.
+    // std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+  for (int i = 0; i < 256 * 1024; ++i) {
+    uint64_t value = iree_prng_splitmix64_next(&state);
+    xxx += value;
+    if (slow) {
+      for (int j = 0; j < 4; ++j) {
+        value = iree_prng_splitmix64_next(&state);
+        xxx += value;
+      }
+    }
+  }
+}
+
+TEST(ExecutorTest, Any) {
+  IREE_TRACE_SCOPE0("ExecutorTest::Any");
+
+  iree_allocator_t allocator = iree_allocator_system();
+
+  iree_task_topology_t* topology = NULL;
+#if 1
+  IREE_CHECK_OK(iree_task_topology_from_physical_cores(
+      /*max_core_count=*/6, allocator, &topology));
+#elif 0
+  IREE_CHECK_OK(iree_task_topology_from_unique_l2_cache_groups(
+      /*max_group_count=*/6, allocator, &topology));
+#else
+  IREE_CHECK_OK(iree_task_topology_from_group_count(/*group_count=*/6,
+                                                    allocator, &topology));
+#endif
+  iree_task_executor_t* executor = NULL;
+  iree_task_scheduling_mode_t scheduling_mode =
+      IREE_TASK_SCHEDULING_MODE_RESERVED;
+  IREE_CHECK_OK(iree_task_executor_create(scheduling_mode, topology, allocator,
+                                          &executor));
+  iree_task_topology_free(topology);
+
+  //
+  iree_task_scope_t scope_a;
+  iree_task_scope_initialize(iree_make_cstring_view("a"), &scope_a);
+
+  //
+  iree_task_call_t call0;
+  iree_task_call_initialize(
+      &scope_a,
+      iree_task_make_closure(
+          [](uintptr_t user_context, uintptr_t task_context) {
+            IREE_TRACE_SCOPE0("call0");
+            EXPECT_EQ(0, user_context);
+            return iree_ok_status();
+          },
+          0),
+      &call0);
+
+  const uint32_t workgroup_size_0[3] = {256, 1, 1};
+  const uint32_t workgroup_count_0[3] = {32, 4, 2};
+  iree_task_dispatch_t dispatch0;
+  iree_task_dispatch_initialize(
+      &scope_a,
+      iree_task_make_closure(
+          [](uintptr_t user_context, uintptr_t task_context) {
+            IREE_TRACE_SCOPE0("tile0");
+            iree_task_tile_context_t* tile_context =
+                (iree_task_tile_context_t*)task_context;
+            EXPECT_EQ(0, user_context);
+            simulate_work(tile_context);
+            iree_atomic_fetch_add_int32(&tile_context->statistics->reserved, 1,
+                                        iree_memory_order_relaxed);
+            return iree_ok_status();
+          },
+          0),
+      workgroup_size_0, workgroup_count_0, &dispatch0);
+  // dispatch0.header.flags |= IREE_TASK_FLAG_DISPATCH_SLICED;
```

Comment:
Useful to keep these comments, `#if 1`s, etc.?

Reply:
(will clean up this file a bit async of the rest of the review)